enhancements/20230830-managed-dhcp.md

```diff
@@ -0,0 +1,318 @@
+	DNS []string `json:"dns,omitempty"`
+	// +optional
+	DomainName *string `json:"domainName,omitempty"`
+	// +optional
+	DomainSearch []string `json:"domainSearch,omitempty"`
+	// +optional
+	NTP []string `json:"ntp,omitempty"`
+	// +optional
+	LeaseTime *int `json:"leaseTime,omitempty"`
+}
+
+type Pool struct {
+	Start string `json:"start,omitempty"`
+	End   string `json:"end,omitempty"`
+	// +optional
+	Exclude []string `json:"exclude,omitempty"`
+}
+
+type IPPoolStatus struct {
+	LastUpdate metav1.Time `json:"lastUpdate,omitempty"`
+	// +optional
+	IPv4 *IPv4Status `json:"ipv4,omitempty"`
+	// +optional
+	AgentPodRef *PodReference `json:"agentPodRef,omitempty"`
+	// // +optional
+	Conditions []genericcondition.GenericCondition `json:"conditions,omitempty"`
+}
+
+type IPv4Status struct {
+	Allocated map[string]string `json:"allocated,omitempty"`
+	Used      int               `json:"used"`
+	Available int               `json:"available"`
+}
+
+type PodReference struct {
+	Namespace string `json:"namespace,omitempty"`
+	Name      string `json:"name,omitempty"`
+}
+```
+
+**VirtualMachineNetworkConfig** is the essentially the static lease of a individual VM.
+
+```go
+var (
+	Allocated condition.Cond = "Allocated"
+	Disabled  condition.Cond = "Disabled"
+)
+
+type VirtualMachineNetworkConfig struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+	Spec   VirtualMachineNetworkConfigSpec   `json:"spec,omitempty"`
+	Status VirtualMachineNetworkConfigStatus `json:"status,omitempty"`
+}
+
+type VirtualMachineNetworkConfigSpec struct {
+	VMName        string          `json:"vmName,omitempty"`
+	NetworkConfig []NetworkConfig `json:"networkConfig,omitempty"`
+}
+
+type NetworkConfig struct {
+	NetworkName string `json:"networkName,omitempty"`
+	MACAddress  string `json:"macAddress,omitempty"`
+	// +optional
+	IPAddress *string `json:"ipAddress,omitempty"`
+}
+
+type VirtualMachineNetworkConfigStatus struct {
+	NetworkConfig []NetworkConfigStatus `json:"networkConfig,omitempty"`
+	Conditions []genericcondition.GenericCondition `json:"conditions,omitempty"`
+}
+
+type NetworkConfigStatus struct {
+	AllocatedIPAddress string `json:"allocatedIPAddress,omitempty"`
+	MACAddress         string `json:"macAddress,omitempty"`
+	NetworkName        string `json:"networkName,omitempty"`
+	Status             string `json:"status,omitempty"`
+}
+```
+
+## Design
+
+Main components:
+
+- Control plane
+  - `vm-dhcp-controller` is majorly responsible for reconciling IPPool and VirtualMachineNetworkConfg objects and maintaining the IPAM cache
+  - `vm-dhcp-webhook` guards the incoming requests for creating, updating, and deleting custom resources
+- Data plane
+  - `vm-dhcp-agent` is scattered in each VM network (defined by NAD), accounting for syncing the pool cache from the IPPool object it monitors and handling the underlying DHCP requests from VMs
+
+![System Architecture](20230830-vm-network-dhcp/system-architecture.svg)
```

Comment:
Please check https://github.com/harvester/harvester/issues/5072#issuecomment-1920653841

The controller local data `IPAllocator` and `CacheAllocator` are anti-k8s-controller patten.